src/mrpro/operators/Functional.py

```diff
@@ -0,0 +1,177 @@
+
+        We assume that functionals are given in the form
+            :math:`f(x) = \phi( weight ( x - target))`
+        for some functional :math:`\phi`.
+
+        Parameters
+        ----------
+        weight
+            weighting of the norm (see above)
+        target
+            element to which distance is taken - often data tensor (see above)
+        dim
+            dimension(s) over which norm is calculated.
+            All other dimensions of  `weight ( x - target))` will be treated as batch dimensions.
+        divide_by_n
+            if True, the result is scaled by the number of elements of the dimensions index by `dim` in
+            the tensor `weight ( x - target))`. If true, the norm is thus calculated as the mean,
+            else the sum.
+        keepdim
+            if true, the dimension(s) of the input indexed by dim are mainted and collapsed to singeltons,
+            else they are removed from the result.
+
+        """
+        super().__init__()
+        self.register_buffer('weight', torch.as_tensor(weight))
+        if target is None:
+            target = torch.tensor(0, dtype=torch.float32)
+        self.register_buffer('target', torch.as_tensor(target))
+        if isinstance(dim, int):
+            dim = (dim,)
+        elif isinstance(dim, Sequence):
+            dim = tuple(dim)
+        self.dim = dim
+        self.divide_by_n = divide_by_n
+        self.keepdim = keepdim
+
+    def _divide_by_n(self, x: torch.Tensor, shape: None | Sequence[int]) -> torch.Tensor:
+        """Apply factor for normalization.
+
+        Input is scaled by the number of elements of either the input
+        or optional shape.
+
+        Parameters
+        ----------
+        x
+            input to be scaled.
+        shape
+            input will be divided by the product of these numbers.
+            If None, it will be divided by the number of elements of the input.
+
+        Returns
+        -------
+            new scaled down tensor.
+        """
+        if not self.divide_by_n:
+            return x
+        if shape is None:
+            shape = x.shape
+        if self.dim is not None:
+            size = [shape[i] for i in self.dim]
+        else:
+            size = list(shape)
+        return x / math.prod(size)
+
+
+class ProximableFunctional(Functional, ABC):
+    r"""ProximableFunctional Base Class.
+
+    A proximable functional is a functional :math:`f(x)` that has a prox implementation,
+    i.e. a function that yields :math:`argmin_x sigma f(x) + 1/2 ||x - y||^2`
+    and a prox_convex_conjugate, yielding the prox of the convex conjugate.
+    """
+
+    @abstractmethod
+    def prox(self, x: torch.Tensor, sigma: torch.Tensor | float = 1.0) -> tuple[torch.Tensor]:
+        r"""Apply proximal operator.
+
+        Yields :math:`prox_{\sigma f}(x) = argmin_{p} (sigma f(p) + 1/2 \|x-p\|^{2}` given `x` and `sigma`
+
+        Parameters
+        ----------
+        x
+            input tensor
```

Comment:
not sure whether this has some impact on the documentation or so... sometimes there is a "." after the description of the input/return parameters, sometimes not.